ext/linux/osq_lock.h

```diff
@@ -0,0 +1,322 @@
+ */
+
+#ifndef __LINUX_OSQ_LOCK_H
+#define __LINUX_OSQ_LOCK_H
+
+#define initialize_lock(lock, threads) osq_lock_init(lock, threads)
+
+#include <stdbool.h>
+#include "atomics.h"
+#include "lk_atomics.h"
+#include "lk_cmpxchg.h"
+#include "lk_barrier.h"
+
+#define ATOMIC_INIT(i)    { (i) }
+
+/*
+ * An MCS like lock especially tailored for optimistic spinning for sleeping
+ * lock implementations (mutex, rwsem, etc).
+ */
+struct optimistic_spin_node {
+    struct optimistic_spin_node *next, *prev;
+    int locked; /* 1 if lock acquired */
+    int cpu; /* encoded CPU # + 1 value */
+};
+
+struct optimistic_spin_queue {
+    /*
+     * Stores an encoded value of the CPU # of the tail node in the queue.
+     * If the queue is empty, then it's set to OSQ_UNLOCKED_VAL.
+     */
+    atomic_t tail;
+};
+
+/* 0 means thread unlocked, 1~N represents each individual thread on core 1~N */
+#define OSQ_UNLOCKED_VAL (0)
+
+/* Init macro and function. */
+#define OSQ_LOCK_UNLOCKED { ATOMIC_INIT(OSQ_UNLOCKED_VAL) }
+
+/*
+ * An MCS like lock especially tailored for optimistic spinning for sleeping
+ * lock implementations (mutex, rwsem, etc).
+ *
+ * Using a single mcs node per CPU is safe because sleeping locks should not be
+ * called from interrupt context and we have preemption disabled while
+ * spinning.
+ */
+
+struct optimistic_spin_node * osq_nodepool_ptr;
+
+static inline void osq_lock_init(uint64_t *lock, unsigned long cores)
+{
+    struct optimistic_spin_queue *osq_ptr = (struct optimistic_spin_queue *) lock;
+    atomic_set(&osq_ptr->tail, OSQ_UNLOCKED_VAL);
+    osq_nodepool_ptr = calloc(cores, sizeof(struct optimistic_spin_node));
+    if (osq_nodepool_ptr == NULL) exit(errno);
+}
+
+bool osq_lock(struct optimistic_spin_queue *lock);
+void osq_unlock(struct optimistic_spin_queue *lock);
+
+static inline bool osq_is_locked(struct optimistic_spin_queue *lock)
+{
+    return atomic_read(&lock->tail) != OSQ_UNLOCKED_VAL;
+}
+
+
+/*
+ * We use the value 0 to represent "no CPU", thus the encoded value
+ * will be the CPU number incremented by 1.
+ */
+static inline int encode_cpu(int cpu_nr)
+{
+    return cpu_nr + 1;
+}
+
+static inline int node_to_cpu(struct optimistic_spin_node *node)
+{
+    return node->cpu - 1;
+}
+
+static inline struct optimistic_spin_node * cpu_to_node(int encoded_cpu_val)
+{
+    int cpu_nr = encoded_cpu_val - 1;
+    return osq_nodepool_ptr + cpu_nr;
+}
+
+/*
+ * Get a stable @node->next pointer, either for unlock() or unqueue() purposes.
+ * Can return NULL in case we were the last queued and we updated @lock instead.
+ */
+static inline struct optimistic_spin_node *
+osq_wait_next(struct optimistic_spin_queue *lock,
+          struct optimistic_spin_node *node,
+          struct optimistic_spin_node *prev)
+{
+    /* sched_getcpu() return current cpu number */
+    int cpu_number = sched_getcpu();
```

Comment:
I am not sure how safe this is, I know we set thread affinity, but there is still the possibility the kernel ignores us (especially because I do not think we check the return codes).  We might need to use the threadnum assigned in lockhammer explicitly to avoid a potential deadlock.

Reply:
Modified in the latest patch and use threadnum instead of sched_getcpu()